kml/minzoom_quadtree.hpp

```diff
@@ -0,0 +1,173 @@
+#pragma once
+
+#include "kml/types.hpp"
+
+#include "indexer/scales.hpp"
+
+#include "geometry/mercator.hpp"
+#include "geometry/rect2d.hpp"
+
+#include "base/assert.hpp"
+#include "base/bits.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <iterator>
+#include <limits>
+#include <utility>
+#include <vector>
+
+namespace kml
+{
+// Builds linear quadtree of input values using coordinates provided
+// then walks through the quadtree elevating minZoom of those values
+// which are greater (in sense of Less) than others at the same
+// quadtree level.
+// To encode each quadtree level uses two bits. One from ordinate and
+// another from abscissa. Both components are mapped to a coordinate
+// system that uses the full range of uint32_t inside the bounding
+// (squared) box.
+template <typename Value, typename Less>
+class MinZoomQuadtree
+{
+public:
+  MinZoomQuadtree(Less const & less) : m_less{less} {}
+
+  template <typename V>
+  void Add(m2::PointD const & point, V && value) { m_quadtree.emplace_back(point, std::forward<V>(value)); }
+
+  void Clear() { m_quadtree.clear(); }
+
+  template <typename F>
+  void SetMinZoom(double countPerTile, int maxZoom, F setMinZoom /* void (Value & value, int minZoom) */)
+  {
+    CHECK_GREATER(countPerTile, 0.0, ());
+    CHECK_LESS_OR_EQUAL(maxZoom, scales::GetUpperStyleScale(), ());
+    CHECK_GREATER_OR_EQUAL(maxZoom, 1, ());
+
+    if (m_quadtree.empty())
+      return;
+
+    m2::RectD bbox;
+    for (auto const & elem : m_quadtree)
+      bbox.Add(elem.m_point);
+
+    if (!(bbox.SizeX() > 0.0 || bbox.SizeY() > 0.0))
+      return;
+
+    // `spacing` is inverse of an characteristic interval between values on the map as if they spaced
```

Comment:
Why do you call it "inverse"?

Reply:
Because I think in terms of source scale, not scale after zooming. Would it be clear if I remove this word?

Reply:
Yes: when I read "inverse of an interval" I expect ```~ 1 / dx``` for some dx.